### ISSUE TITLE
mount volumes within ro volumes causes permanent machine mount failure

### ISSUE THREAD

**Issue by stampycode** [docker/docker#25658]
TL;DR - doing this in your `docker-compose.yml`:
```
volumes:
       - ./abcd:/abcd:ro
       - ./defg:/abcd/defg:ro
```
will permanently break volume mounts in your current docker-machine and all subsequently created docker-machines on the same host, until the host is rebooted.

**Output of `docker version`:**
```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:44:17 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**
```
Containers: 3
 Running: 1
 Paused: 0
 Stopped: 2
Images: 36
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 59
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Kernel Version: 4.4.16-boot2docker
Operating System: Boot2Docker 1.12.0 (TCL 7.2); HEAD : e030bab - Fri Jul 29 00:29:14 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9 MiB
Name: newm
ID: ZWTS:ZUF2:VZOL:IDVC:WUWH:J3IK:GCR7:JT4G:QDTV:IYG4:RBBQ:ZC4R
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug mode (client): false
Debug mode (server): true
 File Descriptors: 24
 Goroutines: 43
 System Time: 2016-08-12T10:27:29.323758623Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running docker on OSX 10.11.5 locally.

**Steps to reproduce the issue:**
1 - create `docker-compose.yml`
```
version: '2'
services:
  crashtestdummy:
     container_name: testcont
     volumes:
       - ./abcd:/abcd:ro
       - ./defg:/abcd/defg:ro
       #  ^^^ this line is the problem, mounting within a ro mount
     image: nginx
```
2 - create test data
```
mkdir abcd && touch abcd/foo
mkdir defg && touch defg/bar
```
3 - build
```
Pulling crashtestdummy (nginx:latest)...
latest: Pulling from library/nginx
51f5c6a04d83: Pull complete
a3ed95caeb02: Pull complete
51d229e136d0: Pull complete
bcd41daec8cc: Pull complete
Digest: sha256:0fe6413f3e30fcc5920bc8fa769280975b10b1c26721de956e1428b9e2f29d04
Status: Downloaded newer image for nginx:latest
Creating testcont

ERROR: for crashtestdummy  oci runtime error: rootfs_linux.go:53: mounting "/mnt/sda1/var/lib/docker/aufs/mnt/db0db52cc8e9b82a60cef552707ffe6e37f738f1fe57efaf87a95cac618edc60/abcd/defg" to rootfs "/mnt/sda1/var/lib/docker/aufs/mnt/db0db52cc8e9b82a60cef552707ffe6e37f738f1fe57efaf87a95cac618edc60" caused "mkdir /mnt/sda1/var/lib/docker/aufs/mnt/db0db52cc8e9b82a60cef552707ffe6e37f738f1fe57efaf87a95cac618edc60/abcd/defg: read-only file system"
Attaching to
```
4 - create a new machine
```
$ docker-machine create --driver virtualbox new4
$ eval $(docker-machine env new4)
```
5 - fix `docker-compose.yml`
```
version: '2'
services:
  crashtestdummy:
     container_name: testcont
     volumes:
       - ./abcd:/abcd:ro
       - ./defg:/defg:ro
     image: nginx 
```
6 - build
```
Pulling crashtestdummy (nginx:latest)...
latest: Pulling from library/nginx
51f5c6a04d83: Pull complete
a3ed95caeb02: Pull complete
51d229e136d0: Pull complete
bcd41daec8cc: Pull complete
Digest: sha256:0fe6413f3e30fcc5920bc8fa769280975b10b1c26721de956e1428b9e2f29d04
Status: Downloaded newer image for nginx:latest
Creating testcont
```
7 - attach and inspect the mounts
```
docker exec -it testcont bash
root@c84386f82b63:/# ls -l /abcd
total 0
root@c84386f82b63:/# ls -l /defg
total 0
root@c84386f82b63:/#
```

**Describe the results you received:**
All mount points are now broken for this docker machine (and all docker machines created after this point) until the host system is rebooted.
Existing docker machines created before this point continue to work normally.
If I fix the docker-compose.yml file, create a new docker machine and repeat the above steps, there are no errors about mount points failing, they just aren't mounted at all.

**Describe the results you expected:**
The second mount point described in the docker file should be broken, the other should be fine.
Resolving the issue in the docker-compose.yml file should fix the problem when the container is rebuilt.

**Additional information you deem important (e.g. issue happens only occasionally):**
This is reproducible every time for me.

docker-inspect on the second machine shows the mounts should be fine, no errors were reported, but the volumes are simply not mounted.
```
"Mounts": [
            {
                "Source": "/data/docker-crash-test/defg",
                "Destination": "/defg",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/data/docker-crash-test/abcd",
                "Destination": "/abcd",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            }
        ],
```

**Comment by stampycode** [comment]
This might actually be a security issue, if a user can knock out the ability for all subsequently created machines to mount volumes... could cause a denial of service for machines the user wouldn't otherwise have access to. Maybe.

**Comment by thaJeztah** [comment]
This looks like an expected issue; you're bind-mounting `abcd` at `/abcd` inside the container, but making it read-only. The `defg` subsequently tries to create a mount at `/abcd/defg`, however, because `/abcd/` is read-only inside the container, it's not possible to create a mount-point there, so it fails.

**Comment by thaJeztah** [comment]
I don't see the reproduction steps for;

> All mount points are now broken for this docker machine (and all docker machines created after this point) until the host system is rebooted.

could you include that information?

**Comment by thaJeztah** [comment]
@stampycode I am not able to reproduce, but I see an error in your report;

```bash
mkdir abcd && touch foo
mkdir defg && touch bar
```

You're creating `foo` and `bar` *next to* the directories `abcd` and `defg`, not inside those directories;

```bash
tree
.
├── abcd
├── bar
├── defg
└── foo

2 directories, 2 files
```

Changing this to;

```bash
mkdir abcd && touch abcd/foo
mkdir defg && touch defg/bar
```

Makes those files inside the directory;

```bash
tree
.
├── abcd
│   └── foo
└── defg
    └── bar

2 directories, 2 files
```

Then, trying to reproduce:

```bash
docker run --rm --name=should-fail -v $PWD/abcd:/abcd:ro -v $PWD/defg:/abcd/defg:ro alpine sh -c 'ls -la /abcd && ls -la /abcd/defg'

docker: Error response from daemon: oci runtime error: rootfs_linux.go:53: mounting "/mnt/sda1/var/lib/docker/aufs/mnt/10a42a2de16f355b22b7c3ae70f443b25133dbeb6cb0bcc7e287eb43c84ede8d/abcd/defg" to rootfs "/mnt/sda1/var/lib/docker/aufs/mnt/10a42a2de16f355b22b7c3ae70f443b25133dbeb6cb0bcc7e287eb43c84ede8d" caused "mkdir /mnt/sda1/var/lib/docker/aufs/mnt/10a42a2de16f355b22b7c3ae70f443b25133dbeb6cb0bcc7e287eb43c84ede8d/abcd/defg: read-only file system".

docker run --rm --name=should-not-fail -v $PWD/abcd:/abcd:ro -v $PWD/defg:/defg:ro alpine sh -c 'ls -la /abcd && ls -la /defg'
total 4
drwxr-xr-x    1 1000     50             102 Aug 12 11:50 .
drwxr-xr-x   26 root     root          4096 Aug 12 11:54 ..
-rw-r--r--    1 1000     50               0 Aug 12 11:50 foo
total 4
drwxr-xr-x    1 1000     50             102 Aug 12 11:50 .
drwxr-xr-x   26 root     root          4096 Aug 12 11:54 ..
-rw-r--r--    1 1000     50               0 Aug 12 11:50 bar
```

Second container runs well, and shows that the bind-mounts are correctly mounted, and contents is there

**Comment by stampycode** [comment]
@thaJeztah  - sorry, that was an error in my description, have updated.

**Comment by stampycode** [comment]
my output from your cmd in a new machine:
```
$ docker run --rm --name=should-not-fail -v $PWD/abcd:/abcd:ro -v $PWD/defg:/defg:ro alpine sh -c 'ls -la /abcd && ls -la /defg'
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
e110a4a17941: Pull complete
Digest: sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a
Status: Downloaded newer image for alpine:latest
total 4
drwxr-xr-x    2 root     root            40 Aug 12 11:14 .
drwxr-xr-x   26 root     root          4096 Aug 12 12:27 ..
total 4
drwxr-xr-x    2 root     root            40 Aug 12 11:14 .
drwxr-xr-x   26 root     root          4096 Aug 12 12:27 ..
```

**Comment by thaJeztah** [comment]
What directory are you running that from (on your OS X host)? Is it a subdirectory of `/Users`? Because only `/Users` is shared with the VirtualBox machine

**Comment by stampycode** [comment]
I am running this in a subdir of `/data/` - which has worked fine up until I discovered this problem.
Honestly hadn't thought that might be an issue.

**Comment by stampycode** [comment]
interestingly, it starts working again if I copy and run from /Users:
```
$ docker run --rm --name=should-not-fail -v $PWD/abcd:/abcd:ro -v $PWD/defg:/defg:ro alpine sh -c 'ls -la /abcd && ls -la /defg'
total 4
drwxr-xr-x    1 1000     50             102 Aug 12 12:32 .
drwxr-xr-x   26 root     root          4096 Aug 12 12:32 ..
-rw-r--r--    1 1000     50               0 Aug 12 12:32 foo
total 4
drwxr-xr-x    1 1000     50             102 Aug 12 12:32 .
drwxr-xr-x   26 root     root          4096 Aug 12 12:32 ..
-rw-r--r--    1 1000     50               0 Aug 12 12:32 foo
```

**Comment by thaJeztah** [comment]
If you bind-mount from another location, docker creates an empty directory on the VirtualBox VM's filesystem, and bind-mounts that directory. If you once had a container that used that location, and wrote files to it, those files are bind-mounted the next time, so perhaps it looked like it worked, but only because the "old" files were present in the VM. Spinning up a new Machine would not have those files, so you'd see an empty directory.

Note that Docker for Mac (https://www.docker.com/products/docker#/mac) is a lot more flexible in this area, and solves issues like permissions, and inotify not working in VirtualBox; you can also share "any" directory on your host, by adding paths;

<img width="380" alt="screen shot 2016-08-12 at 14 50 59" src="https://cloud.githubusercontent.com/assets/1804568/17622728/5fb52306-609c-11e6-9b39-9fbfaf280702.png">

**Comment by stampycode** [comment]
... now I can't get mounts working outside of `/Users` at all... it was working this morning...
*le sigh*
going to install Docker for Mac.

**Comment by cpuguy83** [comment]
The error is pretty plain here:

1. It's creating a missing (or not missing, either way doesn't matter here) `/abcd` dir and then mounting the host path onto it, with readonly
2. It's attempting to create a missing `/abcd/defg` in `/abcd` and failing because `/abcd` is readonly.

This could probably be fixed (in github.com/opencontainers/runc).. but also curious why you don't put `./defg` in `./abcd` to begin with.

It's not a security issue since the user already has access to docker. If a user has access to Docker they have root access to your system.

**Comment by stampycode** [comment]
@cpuguy83 The reasoning behind not putting `./defg` inside `./abcd` to begin with, is because this is simulating putting `defg` inside the `/vendor` dir of my application, where both `abcd` and `defg` are components only available locally - I don't keep one inside the other to reduce the risk of deleting the dependency through an aggressive `composer install` or similar, and because there are multiple applications that depend on this dependency.
I would normally have solved this with symlinking the `defg` dir within the vendors dir in all the local applications - but following symlinks outside of the base docker container dir wasn't supported last time I checked.
This is a local development 'hack' that I use to save myself having to copy the dependency dir to multiple locations.

I'm not aware of how Docker is used securely from a sandboxing perspective, but I would expect that if a server is hosting a docker machine that I have access to via shared key, cert, etc. that I don't automatically have root access to the host server.... and that I wouldn't be automatically granted access to other machines on the same server if I don't have the certs/keys for them - so if I were able to crash my own docker-machine and cause all other docker-machines on the same host to start to fail, would that not constitute a DoS security issue?

**Comment by cpuguy83** [comment]
@stampycode Any person with access to the docker socket has full root access to the host.

**Comment by thaJeztah** [comment]
@stampycode access to the docker socket is equal to having root access on the host, for example:

```
docker run -it --privileged --pid=host debian nsenter -t 1 -m -n sh
```

gives you a root shell on the host

**Comment by stampycode** [comment]
did not know that.
Thanks guys :)

**Comment by teohhanhui** [comment]
> The error is pretty plain here:
>
> 1. It's creating a missing (or not missing, either way doesn't matter here) /abcd dir and then mounting the host path onto it, with readonly
> 2. It's attempting to create a missing /abcd/defg in /abcd and failing because /abcd is readonly.

In my case I'm mounting the project directory as read-only, and trying to mount certain subdirectories as read-write has failed. Are there any plans to fix this?

**Comment by thaJeztah** [comment]
@teohhanhui are the subdirectories themselves already present in the "project" directory, or are they not present, and docker is trying to _create_ a directory as the mount point?

**Comment by teohhanhui** [comment]
They do not exist, but these are directories that are only supposed to exist within the container. And that's why the project directory is mounted (from host) as read-only.

**Comment by thaJeztah** [comment]
So that's the issue; docker cannot create a directory to use as mount point to mount the directory from the host there. Create (empty) directories for those paths in your project and it probably works.

It won't be fixed as it's not a bug
